lib/u3d/downloader.rb

```diff
@@ -36,72 +36,94 @@ module Downloader
+          validator = LinuxValidator.new
           downloader = Downloader::LinuxDownloader
-          files << ["Unity #{version}", downloader.download(version, versions), {}]
+        when :mac
+          validator = MacValidator.new
+          downloader = Downloader::MacDownloader
+        when :win
+          validator = WindowsValidator.new
+          downloader = Downloader::WindowsDownloader
         else
-          downloader = Downloader::MacDownloader if os == :mac
-          downloader = Downloader::WindowsDownloader if os == :win
-          if packages.count.zero?
-            files = downloader.download_all(version, versions)
-          else
-            packages.each do |package|
-              result = downloader.download_specific(package, version, versions)
-              files << [package, result[0], result[1]] unless result.nil?
+          raise ArgumentError, "Operating system #{definition.os.id2name} is not recognized"
+        end
+
+        packages.each do |package|
+          path = downloader.destination_for(package, definition)
+          if File.file?(path)
+            if validator.validate(package, path, definition)
+              files << [package, path, definition[package]]
+            else
+              UI.important "File present at #{path} is not correct, will not be used. Skipping #{package}"
             end
+          else
+            UI.error "No file has been downloaded for #{package}, or it has been moved from #{path}"
           end
         end
+
         files
       end
 
-      # find already downloaded packages or all if none specified
-      def local_files(version, os, packages: [])
-        files = []
-        if os == :linux
-          downloader = Downloader::LinuxDownloader
-          files << ["Unity #{version}", downloader.local_file(version), {}]
-        else
-          downloader = Downloader::MacDownloader if os == :mac
-          downloader = Downloader::WindowsDownloader if os == :win
-          if packages.count.zero?
-            files = downloader.all_local_files(version)
+      private #-----------------------------------------------------------------
+
+      def get_package(downloader, validator, package, definition, files)
+        path = downloader.destination_for(package, definition)
+        url = downloader.url_for(package, definition)
+        if File.file?(path)
+          UI.verbose "Installer file for #{package} seems to be present at #{path}"
+          if validator.validate(package, path, definition)
+            UI.message "#{package.capitalize} is already downloaded"
+            files << [package, path, definition[package]]
+            return
           else
-            packages.each do |package|
-              result = downloader.local_file(package, version)
-              files << [package, result[0], result[1]] unless result.nil?
-            end
+            UI.important "File present at #{path} is not correct, deleting it"
+            File.delete path
```

Comment:
we should add a comment here that we want to be really careful here. Maybe optional not delete it or rename it to something different. I am thinking of non connected installs for CI centers.